app/src/main/java/com/kabouzeid/gramophone/service/MusicService.java

```diff
@@ -1193,9 +1200,16 @@ public void handleMessage(@NonNull final Message msg) {
                     break;
 
                 case TRACK_ENDED:
-                    if (service.getRepeatMode() == REPEAT_MODE_NONE && service.isLastTrack()) {
+                    // if there is a timer finished, don't continue
+                    if (service.pendingQuit ||
+                            service.getRepeatMode() == REPEAT_MODE_NONE && service.isLastTrack()) {
                         service.notifyChange(PLAY_STATE_CHANGED);
                         service.seek(0);
+                        if (service.pendingQuit) {
+                            service.pendingQuit = false;
```

Comment:
Could you potentially just dispatch the actual `ACTION_QUIT` action here to reduce duplicate logic?

Reply:
I did some investigatioin but still don't know how to dispatch service action from itself. Is there a way to acheive this?

Reply:
Hmm, nevermind then if there's no good way of doing it.